### PR TITLE
New interface for Decoder!

### PR DESCRIPTION
**Pull request by swanandx**
The struct `Decoder<T: Crack>` takes a generic argument which implements `Crack` trait.

for creating new decoders, e.g. base85 we can just do it like:

```rust
pub struct Base85Decoder;

impl Base85Decoder {
    pub fn new() -> Decoder<Base85Decoder> {
        Decoder {
                decoder: Base85Decoder,
                // ..other_info here
            }
    }
}

impl Crack for Base85Decoder {
    fn crack(&self, text: &str) -> Option<String> {
        // crack here
    }
}
```

### CODE REVIEW

File: src/checkers/lemmeknow.rs [hunk 1]
```diff
@@ -0,0 +1,7 @@
+use lemmeknow::Identify;
+
+pub fn check_lemmeknow(input: &str) -> bool {
+    // Uses lemmeknow to check if any regexes match
+    let identifier = Identify::default();
```
Comment (bee-san) [inline src/checkers/lemmeknow.rs:5]: Would it be faster to move this out of the function so we don't initialise it everytime?

File: src/checkers/english.rs [hunk 2]
```diff
@@ -0,0 +1,24 @@
+// import storage
+use crate::storage;
+
+// given an input, check every item in the array and return true if any of them match
+pub fn check_english(input: &str) -> Option<&str> {
+    if let Some(result) = storage::DICTIONARIES
+        .iter()
+        .find(|(_, words)| words.contains(input))
```
Comment (bee-san) [inline src/checkers/english.rs:8]: How can we get the word found in the dictionary + the filename from this? :)

File: src/checkers/mod.rs [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+pub fn check(input: &str) -> bool {
+    // Uses lemmeknow to check if any regexes match
+    // import and call lemmeknow.rs
+    if lemmeknow::check_lemmeknow(input) {
+        return true;
+    }
+
+    if english::check_english(input).is_some() {
+        return true;
+    }
```
Comment (bee-san) [inline src/checkers/mod.rs:13]: We should make them return the same thing

File: src/decoders/reverse_decoder.rs [hunk 4]
```diff
@@ -8,29 +8,22 @@ use super::interface::Decoder;
...
-                expected_success: 1.0,
-                failure_runtime: 0.01,
-                normalised_entropy: vec![1.0, 10.0],
-                // I have never seen a reversed string in a CTF
-                // or otherwise
-                popularity: 0.2,
-            },
+    pub fn new() -> Decoder<ReverseDecoder> {
+        Decoder {
+            decoder: ReverseDecoder,
```
Comment (bee-san) [inline src/decoders/reverse_decoder.rs:16]: How do does this work? How do we get the name of the decoder as a string?

**Review by bee-san — CHANGES_REQUESTED** [review]

File: src/checkers/lemmeknow.rs (hunk 1 above)
Comment (swanandx) [reply src/checkers/lemmeknow.rs:5]: Yes! will do it.

File: src/checkers/english.rs (hunk 2 above)
Comment (swanandx) [reply src/checkers/english.rs:8]: do you mean word + filename? as the filename is the name of dictionary right 😁

File: src/decoders/reverse_decoder.rs (hunk 4 above)
Comment (swanandx) [reply src/decoders/reverse_decoder.rs:16]: i thought name was required just to identify the decoder and that can also be done by struct. If we want name, we can just add a `name` field as before.

**Pull request closed without merging**